File: Leetcode2606. Find substring with max cost/Solution.java
class Solution {
    public int maximumCostSubstring(String s, String chars, int[] vals) {
        // build value map for each character
        int val[] = new int[26];
        for(int i=0; i<26; i++) {
            val[i] = i+1;
        }
        // override with custom values from chars and vals
        for(int i=0; i<chars.length(); i++) {
            val[chars.charAt(i)-'a'] = vals[i];
        }

        // kadane's algorithm to find max subarray
        int max=0;
        int n = s.length();
        int cur_sum = 0;
        for(char c:s.toCharArray()) {
            int value = val[c-'a'];
            cur_sum = Math.max(0, cur_sum+value);
            max = Math.max(max, cur_sum);
        }

        return max;
    }
}

// runtime - 3ms
class Solution {
    public int maximumCostSubstring(String s, String chars, int[] vals) {
        int[] map = new int[26];

        for(int i = 0; i < map.length; i++){
            map[i] = i + 1;
        }

        for(int i = 0; i < chars.length(); i++){
            map[chars.charAt(i) - 'a'] = vals[i];
        }

        int result = 0;
        int curr = 0;

        for(int i = 0; i < s.length(); i++){
            if(curr < 0)
                curr = 0;
            
            curr += map[s.charAt(i) - 'a'];
            result = Math.max(result, curr);
        }

        return result;
    }
}


// runtime - 5ms
class Solution {
    public int maximumCostSubstring(String s, String chars, int[] vals) {
        int[] val = new int[26];
        Arrays.fill(val, Integer.MIN_VALUE);

        for (int i = 0; i < chars.length(); i++) {
            val[chars.charAt(i) - 'a'] = vals[i];
        }

        int prefixSum = 0;
        int minPrefix = 0;
        int res = 0;

        for (char c : s.toCharArray()) {
            int v = val[c - 'a'] == Integer.MIN_VALUE
                    ? c - 'a' + 1
                    : val[c - 'a'];

            prefixSum += v;

            res = Math.max(res, prefixSum - minPrefix);
            minPrefix = Math.min(minPrefix, prefixSum);
        }

        return res;
    }
}


// runtime - 6ms
class Solution {
    public int maximumCostSubstring(String s, String chars, int[] vals) {
        int n = s.length();
        int [] value = new int[26];

        Arrays.fill(value, -1001);

        for(int i = 0 ; i < vals.length ; i++){
            value[chars.charAt(i) - 'a'] = vals[i];
        }

        int sum = 0, max = 0;

        for(int i = 0 ; i < n ; i++){
            char ch = s.charAt(i);
            if(value[ch - 'a'] != -1001){
                sum += value[ch - 'a'];
            }else{
                sum += (ch - 'a' + 1);
            }

            if(sum < 0){
                sum = 0;
            }

            if(sum > max){
                max = sum;
            }
        }

        return max;
    }
}


// runtime - 9ms
class Solution {
    public int maximumCostSubstring(String s, String chars, int[] vals) {
        int[] map = new int[26]; 
        Arrays.fill(map,-1001);
        for(int i = 0;i<chars.length();i++){
            map[chars.charAt(i) - 'a'] = vals[i];
        }

        int res = 0;
        int max = 0;
        for(int i = 0;i<s.length();i++){
            int val = (s.charAt(i) - 'a')+1;
            if(map[s.charAt(i) - 'a'] != -1001) {
                val = map[s.charAt(i) - 'a'];
            }
            res+=val;
            if(res<0) res = 0;
            max = Math.max(res,max);
        }
        return max;
    }
}

File: Leetcode2606. Find substring with max cost/Solution.cpp
class Solution {
public:
    int maximumCostSubstring(string s, string chars, vector<int>& vals) {
        // vector v containing values of all the alphabets.
        vector<int> v(26);
        for(int i=0; i<26; i++) v[i]=i+1;
        for(int i=0; i<chars.size(); i++){
            v[chars[i]-'a'] = vals[i];
        }
        int n = s.size();
        // vector corresponding to string s
        vector<int> vt(n);
        for(int i=0; i<s.size(); i++){
            vt[i] = v[s[i]-'a'];
        }
        int ans=0, temp=0;
        for(int i=0; i<n; i++){
            temp += vt[i];
            if(temp<0) temp=0;
            // assign maximum value to ans
            ans = max(ans, temp);
        }
        return ans;
    }
};

// runtime - 0ms
class Solution {
public:
    int maximumCostSubstring(string s, string chars, vector<int>& vals) {
        // Build value map for each character
        vector<int> charValue(26);
        for (int i = 0; i < 26; i++) {
            charValue[i] = i + 1; // default value is 1-indexed position
        }
        
        // Override with custom values from chars and vals
        for (int i = 0; i < chars.length(); i++) {
            charValue[chars[i] - 'a'] = vals[i];
        }
        
        // Kadane's algorithm to find maximum subarray sum
        int maxCost = 0; // empty string has cost 0
        int currentSum = 0;
        
        for (char c : s) {
            int value = charValue[c - 'a'];
            currentSum = max(0, currentSum + value);
            maxCost = max(maxCost, currentSum);
        }
        
        return maxCost;
    }
};

// runtime - 1ms
class Solution {
public:
    int maximumCostSubstring(string s, string chars, vector<int>& vals) {
        int value[26];
        for(int i=0;i<26;i++){
            value[i]=i+1;
        }
        for(int i=0;i<chars.size();i++){
            value[chars[i]-'a']=vals[i];

        }
        int maxSum=0,currSum=0;
        for(int i=0;i<s.size();i++){
            currSum+=value[s[i]-'a'];
            if(currSum<0) currSum=0;
            maxSum=max(maxSum,currSum);
        }
        return maxSum;
        
    }
};


// runtime - 2ms
class Solution {
public:
    int maximumCostSubstring(string s, string chars, vector<int>& vals) {

        int n = s.length();

        vector<int> scores(n) , freq(26 , 1002);

        for(int i=0;i<chars.length();i++){

            freq[chars[i] - 'a'] = vals[i];
        }

        for(int i=0;i<26;i++){
            if(freq[i] == 1002)
                freq[i] = i+1;
        }

        int ans = 0 , curr = 0;

        for(int i=0;i<n;i++){
            scores[i] = freq[s[i] - 'a'];

            curr += scores[i];

            ans = max(ans , curr);

            if(curr < 0){
                curr = 0;
            }
        }

        // int ans = 0 , curr = 0;

        // for(int i=0;i<n;i++){

        //     curr += scores[i];

        //     ans = max(ans , curr);

        //     if(curr < 0){
        //         curr = 0;
        //     }
        // }

        return ans;
    }
};


// runtime - 3ms
class Solution {
public:
    int maximumCostSubstring(string s, string chars, vector<int>& vals) {
        vector<int> costs(26);
        iota(costs.begin(), costs.end(), 1);
        for (int i = 0; i < chars.size(); i++) {
            costs[chars[i] - 'a'] = vals[i];
        }
        //dp[i] = max(dp[i-1] + costs[i], costs[i])
        int ans = 0;
        int dp = 0;
        for (int i = 0; i < s.size(); i++) {
            dp = max(dp + costs[s[i] - 'a'], costs[s[i] - 'a']);
            ans = max(ans, dp);
        }
        return ans;
    }
};


// runtime - 4ms
class Solution {
public:
    int maximumCostSubstring(string s, string chars, vector<int>& vals) {
        
        int cost[26];

        for(char c = 'a'; c <= 'z'; ++c)
        {
            size_t index = chars.find(c);
            if(index == string::npos)
                cost[c - 'a'] = c - 'a' + 1;
            else
                cost[c - 'a'] = vals[index];
        }
        
        int maxCost = 0;
        int currCost = 0;
        for(char &c : s)
        {
            currCost += cost[c - 'a'];
            maxCost = max(maxCost, currCost);
            if(currCost < 0)
                currCost = 0;
        }

        return maxCost;
    }
};


